src/screen-cast.c

```diff
@@ -100,6 +114,142 @@ GType screen_cast_session_get_type (void);
+                                                   NULL,
+                                                   &error))
+    {
+      return NULL;
+    }
+
+  if (!perms || !g_variant_lookup (perms, app_id, "^a&s", &permissions))
+    return NULL;
+
+  return g_steal_pointer (&data);
+}
+
+void
+delete_persistent_permissions (const char *app_id,
+                               const char *restore_token)
+{
+
+  g_autoptr(GError) error = NULL;
+
+  if (!xdp_impl_permission_store_call_delete_sync (get_permission_store (),
+                                                   "screencast",
+                                                   restore_token,
+                                                   NULL,
+                                                   &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_warning ("Error deleting permission: %s", error->message);
+    }
+}
+
+static void
+set_transient_permissions (const char *sender,
+                           const char *restore_token,
+                           GVariant *restore_data)
+{
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&transient_permissions_lock);
+
+  if (!transient_permissions)
+    {
+      transient_permissions = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                                     g_free, (GDestroyNotify)g_variant_unref);
+    }
+
+  g_hash_table_insert (transient_permissions,
+                       g_strdup_printf ("%s:::%s", sender, restore_token),
+                       g_variant_ref (restore_data));
+}
+
+static GVariant *
+get_transient_permissions (const char *sender,
+                           const char *restore_token)
+{
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&transient_permissions_lock);
+  g_autofree char *id = NULL;
+  GVariant *permissions;
+
+  if (!transient_permissions)
+    return NULL;
+
+  id = g_strdup_printf ("%s:::%s", sender, restore_token);
+  permissions = g_hash_table_lookup (transient_permissions, id);
+  return permissions ? g_variant_ref (permissions) : NULL;
+}
+
+static void
+delete_transient_permissions (const char *sender,
+                              const char *restore_token)
+{
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&transient_permissions_lock);
+  g_autofree char *id = NULL;
+
+  if (!transient_permissions)
+    return;
+
+  id = g_strdup_printf ("%s:::%s", sender, restore_token);
+  g_hash_table_remove (transient_permissions, id);
+}
+
+void
+screen_cast_remove_transient_permissions_for_sender (const char *sender)
+{
+  g_autoptr(GMutexLocker) locker = NULL;
+  GHashTableIter iter;
+  const char *key;
+
+  locker = g_mutex_locker_new (&transient_permissions_lock);
+
+  if (!transient_permissions)
+    return;
+
+  g_hash_table_iter_init (&iter, transient_permissions);
+  while (g_hash_table_iter_next (&iter, (gpointer *) &key, NULL))
+    {
+      g_auto(GStrv) split = g_strsplit (key, ":::", 2);
```

Comment:
If the keys in this hash table start with D-Bus bus names, then I'd suggest using a delimiter that is a character that cannot appear in a syntactically valid D-Bus bus name at all (`:` can appear; `:::` cannot, but you have to read the spec more carefully to know that).

`/` would seem like an obvious choice, resulting in something like `:1.42/blah-blah-blah` splitting into sender `:1.42` and transient permissions `blah-blah-blah`.